src/sentry/eventstore/models.py

```diff
@@ -614,11 +563,166 @@ def search_message(self) -> str:
 
         return cast(str, trim(message.strip(), settings.SENTRY_MAX_MESSAGE_LENGTH))
 
-    def __get_column_name(self, column: Column) -> str:
+    def _get_column_name(self, column: Column) -> str:
         # Events are currently populated from the Events dataset
         return cast(str, column.value.event_name)
 
 
+class Event(BaseEvent):
+    def __init__(
+        self,
+        project_id: int,
+        event_id: str,
+        group_id: int | None = None,
+        data: Mapping[str, Any] | None = None,
+        snuba_data: Mapping[str, Any] | None = None,
+        group_ids: Sequence[int] | None = None,
+    ):
+        super().__init__(project_id, event_id, snuba_data=snuba_data)
+        self.group_id = group_id
+        self.group_ids = group_ids
+        self.data = data
+
+    def __getstate__(self) -> Mapping[str, Any]:
+        state = super().__getstate__()
+        state.pop("_group_cache", None)
+        state.pop("_groups_cache", None)
+        return state
+
+    @property
+    def data(self) -> NodeData:
+        return self._data
+
+    @data.setter
+    def data(self, value: Mapping[str, Any]) -> None:
+        node_id = Event.generate_node_id(self.project_id, self.event_id)
+        self._data = NodeData(
+            node_id, data=value, wrapper=EventDict, ref_version=2, ref_func=ref_func
+        )
+
+    @property
+    def group_id(self) -> int | None:
+        # TODO: `group_id` and `group` are deprecated properties on `Event`. We will remove them
+        # going forward. Since events may now be associated with multiple `Group` models, we will
+        # require `GroupEvent` to be passed around. The `group_events` property should be used to
+        # iterate through all `Groups` associated with an `Event`
+        if self._group_id:
+            return self._group_id
+
+        column = self._get_column_name(Columns.GROUP_ID)
+
+        return self._snuba_data.get(column)
+
+    @group_id.setter
+    def group_id(self, value: int | None) -> None:
+        self._group_id = value
+
+    # TODO We need a better way to cache these properties. functools
+    # doesn't quite do the trick as there is a reference bug with unsaved
+    # models. But the current _group_cache thing is also clunky because these
+    # properties need to be stripped out in __getstate__.
+    @property
+    def group(self) -> Group | None:
+        from sentry.models import Group
+
+        if not self.group_id:
+            return None
+        if not hasattr(self, "_group_cache"):
+            self._group_cache = Group.objects.get(id=self.group_id)
+        return self._group_cache
+
+    @group.setter
+    def group(self, group: Group) -> None:
+        self.group_id = group.id
+        self._group_cache = group
+
+    @property
+    def group_ids(self) -> Sequence[int]:
```

Comment:
I think we will very likely have groups in the future which are not directly backed by group IDs (even just for the fact that we probably want hashes to be stored in snuba and resolve them to group ids later).

If we can get away without group ids being a setter and maybe just an iterable, this could make it easier in the future to actually expand and change this.

Reply:
Is this for errors/transactions, or other new events?

It's kind of hard to remove the setter, since we need to assign the group ids to the event for use in eventstream, and also later for use in `post_process_group`. We don't have `_snuba_data` available at these stages. One alternative could be to pass in the actual `Group` objects themselves to a `groups` setter, and fetch the ids out of them in eventstream and post_process_group. Outside of those, we'd just rely on `_snuba_data` like we do currently.

Another option is to just leave `group_ids` in place. It's optional, so it shouldn't prevent us from using other ways to determine `group_id`. So if in the future we were going to determine group from fingerprint that logic could all happen inside of the `group_ids` getter.

Reply:
@mitsuhiko Ok, I've got something that removes `group_ids` entirely and just relies on us passing a list of `Group`s to the `groups` property. We still store `group_ids` internally, since if we pickle the event we need to be able to refetch the groups based on id.